File: src/core/configmanager.cpp
#include "configmanager.h"
#include <QJsonDocument>
#include <QJsonArray>
#include <QJsonObject>
#include <QFile>
#include <QDir>

ConfigManager::ConfigManager(QObject *parent)
    : QObject(parent), m_settings("MOUSART", "MOUSART")
{
    loadQuickCommands();
    loadAutoReplyRules();
    loadSendSequences();
    m_currentProfile = m_settings.value("currentProfile", "default").toString();
}

void ConfigManager::setCurrentProfile(const QString &profile)
{
    if (m_currentProfile == profile) return;
    m_currentProfile = profile;
    m_settings.setValue("currentProfile", profile);
    emit currentProfileChanged();
}

QStringList ConfigManager::profiles() const
{
    QStringList result;
    result << "default";
    QString configDir = QDir::homePath() + "/.mousart/profiles";
    QDir dir(configDir);
    if (dir.exists()) {
        for (const QString &entry : dir.entryList(QStringList() << "*.json", QDir::Files)) {
            QString name = entry;
            name.chop(5); // remove .json
            if (name != "default" && !result.contains(name))
                result << name;
        }
    }
    return result;
}

// Quick Commands
void ConfigManager::addQuickCommand(const QString &name, const QString &data, bool hexMode)
{
    QVariantMap cmd;
    cmd["name"] = name;
    cmd["data"] = data;
    cmd["hex"] = hexMode;
    m_quickCommands.append(cmd);
    saveQuickCommands();
    emit quickCommandsChanged();
}

void ConfigManager::removeQuickCommand(int index)
{
    if (index < 0 || index >= m_quickCommands.size()) return;
    m_quickCommands.removeAt(index);
    saveQuickCommands();
    emit quickCommandsChanged();
}

void ConfigManager::updateQuickCommand(int index, const QString &name, const QString &data, bool hexMode)
{
    if (index < 0 || index >= m_quickCommands.size()) return;
    QVariantMap cmd;
    cmd["name"] = name;
    cmd["data"] = data;
    cmd["hex"] = hexMode;
    m_quickCommands[index] = cmd;
    saveQuickCommands();
    emit quickCommandsChanged();
}

void ConfigManager::moveQuickCommand(int from, int to)
{
    if (from < 0 || from >= m_quickCommands.size()) return;
    if (to < 0 || to >= m_quickCommands.size()) return;
    m_quickCommands.move(from, to);
    saveQuickCommands();
    emit quickCommandsChanged();
}

// Auto-reply Rules
void ConfigManager::addAutoReplyRule(const QString &name, const QString &match, const QString &response, int delay, bool enabled, bool useRegex)
{
    QVariantMap rule;
    rule["name"] = name;
    rule["match"] = match;
    rule["response"] = response;
    rule["delay"] = delay;
    rule["enabled"] = enabled;
    rule["regex"] = useRegex;
    m_autoReplyRules.append(rule);
    saveAutoReplyRules();
    emit autoReplyRulesChanged();
}

void ConfigManager::removeAutoReplyRule(int index)
{
    if (index < 0 || index >= m_autoReplyRules.size()) return;
    m_autoReplyRules.removeAt(index);
    saveAutoReplyRules();
    emit autoReplyRulesChanged();
}

void ConfigManager::updateAutoReplyRule(int index, const QString &name, const QString &match, const QString &response, int delay, bool enabled, bool useRegex)
{
    if (index < 0 || index >= m_autoReplyRules.size()) return;
    QVariantMap rule;
    rule["name"] = name;
    rule["match"] = match;
    rule["response"] = response;
    rule["delay"] = delay;
    rule["enabled"] = enabled;
    rule["regex"] = useRegex;
    m_autoReplyRules[index] = rule;
    saveAutoReplyRules();
    emit autoReplyRulesChanged();
}

void ConfigManager::toggleAutoReplyRule(int index, bool enabled)
{
    if (index < 0 || index >= m_autoReplyRules.size()) return;
    QVariantMap rule = m_autoReplyRules[index].toMap();
    rule["enabled"] = enabled;
    m_autoReplyRules[index] = rule;
    saveAutoReplyRules();
    emit autoReplyRulesChanged();
}

// Send Sequences
void ConfigManager::addSendSequence(const QString &name, const QVariantList &steps)
{
    QVariantMap seq;
    seq["name"] = name;
    seq["steps"] = steps;
    m_sendSequences.append(seq);
    saveSendSequences();
    emit sendSequencesChanged();
}

void ConfigManager::removeSendSequence(int index)
{
    if (index < 0 || index >= m_sendSequences.size()) return;
    m_sendSequences.removeAt(index);
    saveSendSequences();
    emit sendSequencesChanged();
}

void ConfigManager::updateSendSequence(int index, const QString &name, const QVariantList &steps)
{
    if (index < 0 || index >= m_sendSequences.size()) return;
    QVariantMap seq;
    seq["name"] = name;
    seq["steps"] = steps;
    m_sendSequences[index] = seq;
    saveSendSequences();
    emit sendSequencesChanged();
}

// Profiles
void ConfigManager::saveProfile(const QString &name)
{
    QString configDir = QDir::homePath() + "/.mousart/profiles";
    QDir().mkpath(configDir);
    QString filePath = configDir + "/" + name + ".json";

    QJsonObject profile;
    profile["serialConfig"] = QJsonObject::fromVariantMap(loadSerialConfig());

    QJsonArray cmds;
    for (const auto &cmd : m_quickCommands) {
        cmds.append(QJsonObject::fromVariantMap(cmd.toMap()));
    }
    profile["quickCommands"] = cmds;

    QJsonArray rules;
    for (const auto &rule : m_autoReplyRules) {
        rules.append(QJsonObject::fromVariantMap(rule.toMap()));
    }
    profile["autoReplyRules"] = rules;

    QJsonArray seqs;
    for (const auto &seq : m_sendSequences) {
        seqs.append(QJsonObject::fromVariantMap(seq.toMap()));
    }
    profile["sendSequences"] = seqs;

    QFile file(filePath);
    if (file.open(QIODevice::WriteOnly)) {
        file.write(QJsonDocument(profile).toJson());
        file.close();
    }
    emit profilesChanged();
}

void ConfigManager::loadProfile(const QString &name)
{
    QString configDir = QDir::homePath() + "/.mousart/profiles";
    QString filePath = configDir + "/" + name + ".json";
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) return;

    QJsonDocument doc = QJsonDocument::fromJson(file.readAll());
    file.close();
    QJsonObject profile = doc.object();

    if (profile.contains("serialConfig")) {
        saveSerialConfig(profile["serialConfig"].toObject().toVariantMap());
    }

    if (profile.contains("quickCommands")) {
        m_quickCommands.clear();
        for (const auto &v : profile["quickCommands"].toArray()) {
            m_quickCommands.append(v.toObject().toVariantMap());
        }
        saveQuickCommands();
        emit quickCommandsChanged();
    }

    if (profile.contains("autoReplyRules")) {
        m_autoReplyRules.clear();
        for (const auto &v : profile["autoReplyRules"].toArray()) {
            m_autoReplyRules.append(v.toObject().toVariantMap());
        }
        saveAutoReplyRules();
        emit autoReplyRulesChanged();
    }

    if (profile.contains("sendSequences")) {
        m_sendSequences.clear();
        for (const auto &v : profile["sendSequences"].toArray()) {
            m_sendSequences.append(v.toObject().toVariantMap());
        }
        saveSendSequences();
        emit sendSequencesChanged();
    }

    setCurrentProfile(name);
}

void ConfigManager::deleteProfile(const QString &name)
{
    if (name == "default") return;
    QString configDir = QDir::homePath() + "/.mousart/profiles";
    QFile::remove(configDir + "/" + name + ".json");
    emit profilesChanged();
}

// Serial config persistence
void ConfigManager::saveSerialConfig(const QVariantMap &config)
{
    m_settings.beginGroup("SerialConfig");
    for (auto it = config.begin(); it != config.end(); ++it) {
        m_settings.setValue(it.key(), it.value());
    }
    m_settings.endGroup();
}

QVariantMap ConfigManager::loadSerialConfig()
{
    QVariantMap config;
    m_settings.beginGroup("SerialConfig");
    for (const QString &key : m_settings.childKeys()) {
        config[key] = m_settings.value(key);
    }
    m_settings.endGroup();
    return config;
}

// Private helpers
void ConfigManager::saveQuickCommands()
{
    QJsonArray arr;
    for (const auto &cmd : m_quickCommands) {
        arr.append(QJsonObject::fromVariantMap(cmd.toMap()));
    }
    m_settings.setValue("quickCommands", QJsonDocument(arr).toJson(QJsonDocument::Compact));
}

void ConfigManager::loadQuickCommands()
{
    QByteArray data = m_settings.value("quickCommands").toByteArray();
    if (data.isEmpty()) {
        // Add default commands
        addQuickCommand("AT", "AT\r\n", false);
        addQuickCommand("AT+RST", "AT+RST\r\n", false);
        addQuickCommand("Ping", "FF 01 02 03", true);
        return;
    }
    QJsonDocument doc = QJsonDocument::fromJson(data);
    for (const auto &v : doc.array()) {
        m_quickCommands.append(v.toObject().toVariantMap());
    }
}

void ConfigManager::saveAutoReplyRules()
{
    QJsonArray arr;
    for (const auto &rule : m_autoReplyRules) {
        arr.append(QJsonObject::fromVariantMap(rule.toMap()));
    }
    m_settings.setValue("autoReplyRules", QJsonDocument(arr).toJson(QJsonDocument::Compact));
}

void ConfigManager::loadAutoReplyRules()
{
    QByteArray data = m_settings.value("autoReplyRules").toByteArray();
    if (data.isEmpty()) return;
    QJsonDocument doc = QJsonDocument::fromJson(data);
    for (const auto &v : doc.array()) {
        m_autoReplyRules.append(v.toObject().toVariantMap());
    }
}

void ConfigManager::saveSendSequences()
{
    QJsonArray arr;
    for (const auto &seq : m_sendSequences) {
        arr.append(QJsonObject::fromVariantMap(seq.toMap()));
    }
    m_settings.setValue("sendSequences", QJsonDocument(arr).toJson(QJsonDocument::Compact));
}

void ConfigManager::loadSendSequences()
{
    QByteArray data = m_settings.value("sendSequences").toByteArray();
    if (data.isEmpty()) return;
    QJsonDocument doc = QJsonDocument::fromJson(data);
    for (const auto &v : doc.array()) {
        m_sendSequences.append(v.toObject().toVariantMap());
    }
}

File: mousart/core/config_manager.py
"""Configuration persistence - quick commands, auto-reply rules, profiles."""
import json
import os
from mousart.qt_compat import *

from mousart.utils.constants import DEFAULT_QUICK_COMMANDS, DEFAULT_MAX_LOG_ENTRIES


class ConfigManager(QObject):
    """Manages application configuration, quick commands, and profiles."""

    quick_commands_changed = pyqtSignal()
    auto_reply_rules_changed = pyqtSignal()
    send_sequences_changed = pyqtSignal()
    current_profile_changed = pyqtSignal()
    profiles_changed = pyqtSignal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self._settings = QSettings("MOUSART", "MOUSART")
        self._current_profile = self._settings.value("currentProfile", "default")
        self._quick_commands = []
        self._auto_reply_rules = []
        self._send_sequences = []
        self._load_quick_commands()
        self._load_auto_reply_rules()
        self._load_send_sequences()

    @property
    def quick_commands(self):
        return self._quick_commands

    @property
    def auto_reply_rules(self):
        return self._auto_reply_rules

    @property
    def send_sequences(self):
        return self._send_sequences

    @property
    def current_profile(self):
        return self._current_profile

    @current_profile.setter
    def current_profile(self, profile):
        if self._current_profile == profile:
            return
        self._current_profile = profile
        self._settings.setValue("currentProfile", profile)
        self.current_profile_changed.emit()

    @property
    def max_log_entries(self) -> int:
        val = self._settings.value("maxLogEntries", DEFAULT_MAX_LOG_ENTRIES)
        try:
            return int(val)
        except (TypeError, ValueError):
            return DEFAULT_MAX_LOG_ENTRIES

    @max_log_entries.setter
    def max_log_entries(self, value: int):
        if self.max_log_entries == value:
            return
        self._settings.setValue("maxLogEntries", value)

    def get_profiles(self) -> list:
        result = ["default"]
        config_dir = os.path.join(os.path.expanduser("~"), ".mousart", "profiles")
        if os.path.isdir(config_dir):
            for f in sorted(os.listdir(config_dir)):
                if f.endswith(".json"):
                    name = f[:-5]
                    if name != "default" and name not in result:
                        result.append(name)
        return result

    # Quick Commands
    def addQuickCommand(self, name: str, data: str, hex_mode: bool):
        self._quick_commands.append({"name": name, "data": data, "hex": hex_mode})
        self._save_quick_commands()
        self.quick_commands_changed.emit()

    def removeQuickCommand(self, index: int):
        if 0 <= index < len(self._quick_commands):
            self._quick_commands.pop(index)
            self._save_quick_commands()
            self.quick_commands_changed.emit()

    def updateQuickCommand(self, index: int, name: str, data: str, hex_mode: bool):
        if 0 <= index < len(self._quick_commands):
            self._quick_commands[index] = {"name": name, "data": data, "hex": hex_mode}
            self._save_quick_commands()
            self.quick_commands_changed.emit()

    def moveQuickCommand(self, frm: int, to: int):
        if (0 <= frm < len(self._quick_commands) and 0 <= to < len(self._quick_commands)):
            cmd = self._quick_commands.pop(frm)
            self._quick_commands.insert(to, cmd)
            self._save_quick_commands()
            self.quick_commands_changed.emit()

    # Auto-reply Rules
    def addAutoReplyRule(self, name: str, match: str, response: str,
                         delay: int, enabled: bool, use_regex: bool):
        self._auto_reply_rules.append({
            "name": name, "match": match, "response": response,
            "delay": delay, "enabled": enabled, "regex": use_regex
        })
        self._save_auto_reply_rules()
        self.auto_reply_rules_changed.emit()

    def removeAutoReplyRule(self, index: int):
        if 0 <= index < len(self._auto_reply_rules):
            self._auto_reply_rules.pop(index)
            self._save_auto_reply_rules()
            self.auto_reply_rules_changed.emit()

    def updateAutoReplyRule(self, index: int, name: str, match: str,
                            response: str, delay: int, enabled: bool, use_regex: bool):
        if 0 <= index < len(self._auto_reply_rules):
            self._auto_reply_rules[index] = {
                "name": name, "match": match, "response": response,
                "delay": delay, "enabled": enabled, "regex": use_regex
            }
            self._save_auto_reply_rules()
            self.auto_reply_rules_changed.emit()

    def toggleAutoReplyRule(self, index: int, enabled: bool):
        if 0 <= index < len(self._auto_reply_rules):
            self._auto_reply_rules[index]["enabled"] = enabled
            self._save_auto_reply_rules()
            self.auto_reply_rules_changed.emit()

    # Send Sequences
    def addSendSequence(self, name: str, steps: list):
        self._send_sequences.append({"name": name, "steps": steps})
        self._save_send_sequences()
        self.send_sequences_changed.emit()

    def removeSendSequence(self, index: int):
        if 0 <= index < len(self._send_sequences):
            self._send_sequences.pop(index)
            self._save_send_sequences()
            self.send_sequences_changed.emit()

    def updateSendSequence(self, index: int, name: str, steps: list):
        if 0 <= index < len(self._send_sequences):
            self._send_sequences[index] = {"name": name, "steps": steps}
            self._save_send_sequences()
            self.send_sequences_changed.emit()

    # Profiles
    def saveProfile(self, name: str):
        config_dir = os.path.join(os.path.expanduser("~"), ".mousart", "profiles")
        os.makedirs(config_dir, exist_ok=True)
        path = os.path.join(config_dir, f"{name}.json")

        profile = {
            "serialConfig": self.loadSerialConfig(),
            "quickCommands": self._quick_commands,
            "autoReplyRules": self._auto_reply_rules,
            "sendSequences": self._send_sequences,
        }

        with open(path, "w", encoding="utf-8") as f:
            json.dump(profile, f, ensure_ascii=False, indent=2)
        self.profiles_changed.emit()

    def loadProfile(self, name: str):
        config_dir = os.path.join(os.path.expanduser("~"), ".mousart", "profiles")
        path = os.path.join(config_dir, f"{name}.json")
        if not os.path.isfile(path):
            return

        with open(path, "r", encoding="utf-8") as f:
            profile = json.load(f)

        if "serialConfig" in profile:
            self.saveSerialConfig(profile["serialConfig"])

        if "quickCommands" in profile:
            self._quick_commands = profile["quickCommands"]
            self._save_quick_commands()
            self.quick_commands_changed.emit()

        if "autoReplyRules" in profile:
            self._auto_reply_rules = profile["autoReplyRules"]
            self._save_auto_reply_rules()
            self.auto_reply_rules_changed.emit()

        if "sendSequences" in profile:
            self._send_sequences = profile["sendSequences"]
            self._save_send_sequences()
            self.send_sequences_changed.emit()

        self.current_profile = name

    def deleteProfile(self, name: str):
        if name == "default":
            return
        config_dir = os.path.join(os.path.expanduser("~"), ".mousart", "profiles")
        path = os.path.join(config_dir, f"{name}.json")
        if os.path.isfile(path):
            os.remove(path)
        self.profiles_changed.emit()

    # Serial config persistence
    def saveSerialConfig(self, config: dict):
        self._settings.beginGroup("SerialConfig")
        for key, value in config.items():
            self._settings.setValue(key, value)
        self._settings.endGroup()

    def loadSerialConfig(self) -> dict:
        config = {}
        self._settings.beginGroup("SerialConfig")
        for key in self._settings.childKeys():
            config[key] = self._settings.value(key)
        self._settings.endGroup()
        return config

    # Private helpers
    def _save_quick_commands(self):
        data = json.dumps(self._quick_commands, ensure_ascii=False)
        self._settings.setValue("quickCommands", data)

    def _load_quick_commands(self):
        data = self._settings.value("quickCommands", "")
        if not data:
            self._quick_commands = list(DEFAULT_QUICK_COMMANDS)
            self._save_quick_commands()
            return
        try:
            self._quick_commands = json.loads(data)
        except (json.JSONDecodeError, TypeError):
            self._quick_commands = list(DEFAULT_QUICK_COMMANDS)

    def _save_auto_reply_rules(self):
        data = json.dumps(self._auto_reply_rules, ensure_ascii=False)
        self._settings.setValue("autoReplyRules", data)

    def _load_auto_reply_rules(self):
        data = self._settings.value("autoReplyRules", "")
        if not data:
            return
        try:
            self._auto_reply_rules = json.loads(data)
        except (json.JSONDecodeError, TypeError):
            self._auto_reply_rules = []

    def _save_send_sequences(self):
        data = json.dumps(self._send_sequences, ensure_ascii=False)
        self._settings.setValue("sendSequences", data)

    def _load_send_sequences(self):
        data = self._settings.value("sendSequences", "")
        if not data:
            return
        try:
            self._send_sequences = json.loads(data)
        except (json.JSONDecodeError, TypeError):
            self._send_sequences = []
